File: coding/python/programmers/level1/get_report_result.py
"""
  - id_list의 길이가 N일 때, NxN 크기의 2차원 리스트 arr을 선언
    - i번째 유저가 j번째 유저를 신고한 경우, arr[i][j]에 1을 할당.
  - arr 리스트에서 k번 이상 신고된 유저를 탐색한다.
  - 탐색한 유저를 신고한 이력이 있는 i번째 유저에 대해, answer[i]의 값을 ++ 처리한다.
"""


def solution(id_list, report, k):
    id_list_len = len(id_list)
    answer = [0] * id_list_len
    arr = [[0] * id_list_len for _ in range(0, id_list_len)]  # 2차원 리스트

    for condition in report:
        reporter, respondent = condition.split()  # ["muzi frodo"] => "muzi", "frodo" 로 할당해준다.
        reporter_idx = id_list.index(reporter)
        respondent_idx = id_list.index(respondent)
        arr[reporter_idx][respondent_idx] = 1  # 신고한 유저의 index를 1로 변경해준다.

    for i in range(0, id_list_len):
        count = 0
        for j in range(0, id_list_len):
            if arr[j][i] == 1:
                count += 1

        if count >= k:  # k번 이상 신고당한 사람
            for j in range(0, id_list_len):
                if arr[j][i] == 1:
                    answer[j] += 1

    return answer


if __name__ == "__main__":
    print(solution(["muzi", "frodo", "apeach", "neo"], ["muzi frodo", "apeach frodo", "frodo neo", "muzi neo", "apeach muzi"], 2))
    # solution(["con", "ryan"], ["ryan con", "ryan con", "ryan con", "ryan con"], 3)


"""
프로그래머스 "다른사람의 풀이" 중, 
set()을 사용한 깔끔한 풀이가 있어서 첨부했습니다.

def solution(id_list, report, k):
    answer = [0] * len(id_list)    
    reports = {x : 0 for x in id_list}

    for r in set(report):
        reports[r.split()[1]] += 1

    for r in set(report):
        if reports[r.split()[1]] >= k:
            answer[id_list.index(r.split()[0])] += 1

    return answer
    
"""

File: coding/swift/Problem-Solving/Problem-Solving/Programmers/GetReportResult.swift
import Foundation

func solution(_ id_list:[String], _ report:[String], _ k:Int) -> [Int] {
    var result: [Int] = Array(repeating: 0, count: id_list.count)   // 각 유저가 받은 결과 메일 수
    var reportedInfo = [String: Set<String>]()                      // 유저별 신고당한 정보
    var suspendedUser = [String]()                                  // 정지된 유저 ID
    
    // 딕셔너리 만들기
    id_list.forEach { id in
        reportedInfo[id] = []
    }
    
    // 유저별 신고당한 정보 구하기
    report.map { $0.split(separator: " ") }.forEach { content in
        let reporter = String(content[0])
        let reported = String(content[1])
        reportedInfo[reported]?.insert(reporter)    // 동일 중복 신고 방지
    }
    
    // 이용 정지 유저 구하기
    for report in reportedInfo {
        if report.value.count >= k {
            suspendedUser.append(report.key)
        }
    }
//    reportedInfo.forEach { report in
//        if report.value.count >= k {
//            suspendedUser.append(report.key)
//        }
//    }
    
    // 유저별 받을 결과 메일 수 구하기
    for key in suspendedUser {
        for id in reportedInfo[key]! {
            let idx = id_list.firstIndex(of: id)
            result[idx!] += 1
        }
    }
    
//    suspendedUser.forEach { key in
//        reportedInfo[key]!.forEach { id in
//            print(id)
//            result[id_list.firstIndex(of: id)!] += 1
//        }
//    }
    
    return result
}

// [2,1,1,0]
// print(solution(["muzi", "frodo", "apeach", "neo"], ["muzi frodo","apeach frodo","frodo neo","muzi neo","apeach muzi"], 2))
